### DISCUSSION TITLE
SemanticSegmentationConfig's predict_options vs chip_options?

### DISCUSSION THREAD

**Discussion by NoaMills-USDA** [azavea/raster-vision#2194] [Q&A] [answered]
Could you please help me understand the difference between the `predict_options` field and the `chip_options` field of the `SemanticSegmentationConfig` object? I'd like to specify the chip size, stride length, padding, and pad direction of the chips in the validation and test sets. Within the `SemanticSegmentationChipOptions` object, I can specify the stride length, padding, pad direction, and "size" which appears to be different from chip size because it allows for the resizing of images. Within the `SemanticSegmentationPredictOptions` object, I can specify the stride length and the chip size, but not the padding or pad direction. Do I need to specify some parameters in one object and other parameters in another?

**Comment by NoaMills-USDA** [comment]
To add some more context, here's what my `SemanticSegmentationConfig` object looked like before I added stride and padding:

```
return SemanticSegmentationConfig(
        root_uri=output_uri,
        dataset=scene_dataset,
        backend=backend,
        predict_options=SemanticSegmentationPredictOptions(chip_sz=chip_sz))
```

And here's what it looks like after. My code runs, but I'm wondering what the affect is of having both the `chip_sz` field in `SemanticSegmentationPredictionOptions`, and the `size` field in `SemanticSegmentationChipOptions` both set, and if I have the stride parameter in the appropriate place:

```
return SemanticSegmentationConfig(
    root_uri=output_uri,
    dataset=scene_dataset,
    backend=backend,
    predict_options=SemanticSegmentationPredictOptions(chip_sz=chip_sz),
    chip_options=SemanticSegmentationChipOptions(
        sampling=WindowSamplingConfig(
            padding=padding,
            pad_direction="both",
            stride=stride_length,
            size=chip_sz
        )
    )
)

```

**Comment by AdeelH** [comment, marked as answer]
`chip_options` applies to the `chip` stage (which extracts chips into image files before the `train` stage) of the pipeline. If you are using `GeoDataset`s (via `GeoDataConfig`), then you are not using the `chip` stage, so `chip_options` has no effect. See the effect of the `nochip` param in [this example](https://github.com/azavea/raster-vision/blob/v0.30.1/rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/isprs_potsdam.py).

Similarly, `predict_options` applies to the `predict` stage. It currently does not allow specifying `padding` and `pad_direction`; however, as of #2190 (not released yet), the padding value used by default is calculated such that pixels around the edges of the scene do not get left out of the prediction, even if `crop_sz` is specified. 

It might make sense, in the future, to let `PredictOptions` take a `WindowSamplingConfig` just like `ChipOptions` and issue a warning if the user passes a random sampling config (because that's probably not what you want to do when predicting).
